dali/operators/decoder/nvjpeg/decoupled_api/nvjpeg_decoder_decoupled_api.h

```diff
@@ -35,6 +35,84 @@
 
 namespace dali {
 
+namespace memory {
+
+using Buffer = std::pair<void*, size_t>;
+using BufferPool = std::map<std::thread::id, std::vector<Buffer>>;
+static BufferPool device_buffer_pool_;
+static BufferPool pinned_buffer_pool_;
+
+static int DeviceNew(void **ptr, size_t size) {
+  std::thread::id this_id = std::this_thread::get_id();
+  std::cout << "Requesting device memory for Thread " << this_id << ": " << size << " bytes, ptr = ";
+  auto it = device_buffer_pool_.find(this_id);
+  assert(it != device_buffer_pool_.end());
+  auto &buffers = it->second;
+  auto buf = buffers.back();
+  buffers.pop_back();
+  if (size == buf.second) {
+    *ptr = buf.first;
+  } else {
+    std::cout << "Different size, invoking cudaMalloc\n"; 
+    CUDA_CALL(cudaMalloc(ptr, size));
+  }
+  std::cout << *ptr << "\n";
+  return 0;
+}
+
+static int DeviceDelete(void *ptr) {
+  std::thread::id this_id = std::this_thread::get_id();
+  std::cout << "Thread " << this_id << " deleting device memory: ptr = " << ptr << "\n";
+  CUDA_CALL(cudaFree(ptr));
+  return 0;
+}
+
+static nvjpegDevAllocator_t GetDeviceAllocator() {
+  nvjpegDevAllocator_t allocator;
+  allocator.dev_malloc = &memory::DeviceNew;
+  allocator.dev_free = &memory::DeviceDelete;
+  return allocator;
+}
+
+static int PinnedNew(void** ptr, size_t size, unsigned int flags) {
+  std::thread::id this_id = std::this_thread::get_id();
+  std::cout << "Requesting pinned memory for Thread " << this_id << ": " << size << " bytes, ptr = ";
+  assert(flags == 0);
+  auto it = pinned_buffer_pool_.find(this_id);
+  assert(it != pinned_buffer_pool_.end());
+  auto &buffers = it->second;
+  auto buf = buffers.back();
+  buffers.pop_back();
+  if (size == buf.second) {
+    *ptr = buf.first;
+  } else {
+    std::cout << "Different size, invoking cudaHostAlloc\n"; 
+    CUDA_CALL(cudaHostAlloc(ptr, size, flags));
```

Comment:
The only difference compared to DeviceNew is `cudaHostAlloc` itself.

Reply:
Unified now